columnar/src/columnar/merge/mod.rs

```diff
@@ -174,97 +192,174 @@ fn merge_column(
-    for column in columns.iter().flatten() {
-        let min_value: NumericalValue;
-        let max_value: NumericalValue;
-        match column {
-            DynamicColumn::I64(column) => {
-                min_value = column.min_value().into();
-                max_value = column.max_value().into();
-            }
-            DynamicColumn::U64(column) => {
-                min_value = column.min_value().into();
-                max_value = column.min_value().into();
-            }
-            DynamicColumn::F64(column) => {
-                min_value = column.min_value().into();
-                max_value = column.min_value().into();
-            }
-            DynamicColumn::Bool(_)
-            | DynamicColumn::IpAddr(_)
-            | DynamicColumn::DateTime(_)
-            | DynamicColumn::Bytes(_)
-            | DynamicColumn::Str(_) => {
-                panic!("We expected only numerical columns.");
-            }
+fn coerce_columns(
+    column_type: ColumnType,
+    columns: &mut [Option<DynamicColumn>],
+) -> io::Result<()> {
+    for column_opt in columns.iter_mut() {
+        if let Some(column) = column_opt.take() {
+            *column_opt = Some(coerce_column(column_type, column)?);
         }
-        column_types.insert(column.column_type().numerical_type().unwrap());
-        compatible_numerical_types.accept_value(min_value);
-        compatible_numerical_types.accept_value(max_value);
     }
-    if column_types.len() <= 1 {
-        // No need to do anything. The columns are already all from the same type.
-        // This is necessary to let use force a given type.
+    Ok(())
+}
 
-        // TODO This works in a world where we do not allow a change of schema,
-        // but in the future, we will have to pass some kind of schema to enforce
-        // the logic.
-        return;
-    }
-    let coerce_type = compatible_numerical_types.to_numerical_type();
-    for column_opt in columns.iter_mut() {
-        if let Some(column) = column_opt.take() {
-            *column_opt = column.coerce_numerical(coerce_type);
+fn coerce_column(column_type: ColumnType, column: DynamicColumn) -> io::Result<DynamicColumn> {
+    if let Some(numerical_type) = column_type.numerical_type() {
+        column
+            .coerce_numerical(numerical_type)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, ""))
+    } else {
+        if column.column_type() != column_type {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!(
+                    "Cannot coerce column of type `{:?}` to `{column_type:?}`",
+                    column.column_type()
+                ),
+            ));
         }
+        Ok(column)
+    }
+}
+
+fn min_max_if_numerical(column: &DynamicColumn) -> Option<(NumericalValue, NumericalValue)> {
+    match column {
+        DynamicColumn::I64(column) => Some((column.min_value().into(), column.max_value().into())),
```

Comment:
what happens for empty columns?

Reply:
This is not specific to this PR but I added a comment and a unit test.